### PR TITLE
stylo: Improve restyling performance

### PR DESCRIPTION
**Pull request by emilio**
This commit adds hooks to the Servo style traversal to avoid traversing all the
DOM for every restyle. Additionally it changes the behavior of the dirty flag to
be propagated top down, to prevent extra overhead when an element is dirtied.

This commit doesn't aim to change the behavior on Servo just yet, since Servo does extra job when dirtying the node related with DOM revision counters that might be necessary.

CC @asajeffrey for the DOM revision counters stuff. When a node is dirty, do all its descendants really need to increment the revision counter, or is this an unintended effect? My intuition is that this is hurting performance quite a lot for servo.

r? @bholley 

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

- [x] These changes do not require tests because no geckolib tests yet.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12563)

### CODE REVIEW

**Comment by emilio** [comment]
I *think* these are servo incremental layout bugs uncovered by this patch (since right now we relayout way less than before), I'll verify this during the weekend.

**Comment by emilio** [comment]
What a better way to know?

@bors-servo: try

**Comment by emilio** [comment]
Sorry for the spam, the linux builders failed because there was a typo in the geckolib traversal.rs file, but I wanted to look at the [mac results](http://build.servo.org/builders/mac-rel-css/builds/2187/steps/test/logs/stdio), which are more interesting.

**Comment by emilio** [comment]
@bors-servo: try

I think I've got a few (all?) of them, was a super-stupid thing, layout only checked for `has_dirty_descendants`, which in practice **before** it meant `is_dirty || has_dirty_descendants` due to how the flags were propagated.

**Comment by emilio** [comment]
Ok, seems like all the tests are passing except multi-process ones. Need to figure out how I broke it.

**Comment by emilio** [comment]
@samlh Sure, so as an initial way to measure, I just grabbed a twitter feed page, removed scripts and iframes, and loaded it with a release build of servo before and after this patch, profiling with `-p 5`, navigating normally trying to interact with elements to trigger incremental restyle. 

Unfortunately I'm worried this is not a exact measure (even the tiniest difference in interaction could trigger way different code paths), but it might be a good approx measure if done for enough time (these are about 120 incremental restyle events, which are not that many, but probably enough).

Anyway, in that (unexact, but probably sort of relevant) test, the layout numbers are as follows:

### Before patch

```
_category_                          _incremental?_ _iframe?_             _url_                  _mean (ms)_   _median (ms)_      _min (ms)_      _max (ms)_       _events_ 
Compositing                             N/A          N/A                  N/A                        0.6851          0.1803          0.0626        196.0905             559
Layout                                  yes          yes     file:///home/emilio/projects/m         76.6177         55.2281         28.1311       5692.9606             385
Layout                                  no           yes     file:///home/emilio/projects/m        990.0932        990.0932        990.0932        990.0932               1
+ Style Recalc                          yes          yes     file:///home/emilio/projects/m         11.2728         10.4425          3.5714        101.0940             117
+ Style Recalc                          no           yes     file:///home/emilio/projects/m        840.2206        840.2206        840.2206        840.2206               1
| + Text Shaping                        yes          yes     file:///home/emilio/projects/m          0.0583          0.0046          0.0000          2.2492              93
| + Text Shaping                        no           yes     file:///home/emilio/projects/m          2.7505          2.7505          2.7505          2.7505               1
```

### After patch

```
_category_                          _incremental?_ _iframe?_             _url_                  _mean (ms)_   _median (ms)_      _min (ms)_      _max (ms)_       _events_ 
Compositing                             N/A          N/A                  N/A                        0.4669          0.1659          0.0518        131.3618             799
Layout                                  yes          yes     file:///home/emilio/projects/m         74.4345         55.6957         48.5256       6706.9546             399
Layout                                  no           yes     file:///home/emilio/projects/m       1096.8402       1096.8402       1096.8402       1096.8402               1
+ Style Recalc                          yes          yes     file:///home/emilio/projects/m          5.3830          3.6911          0.9647        182.8923             138
+ Style Recalc                          no           yes     file:///home/emilio/projects/m        840.3075        840.3075        840.3075        840.3075               1
| + Text Shaping                        yes          yes     file:///home/emilio/projects/m          0.0483          0.0053          0.0000          2.7329             115
| + Text Shaping                        no           yes     file:///home/emilio/projects/m          2.4944          2.4944          2.4944          2.4944               1
```

Which is more than a 50% improvement on mean incremental style recalc (5.3s vs 11.2s). Again, while this allows some sort of measurement, take these with a *huge* grain of salt, since these are not exact tests. Also, note that this matters even more for the case I was optimizing (stylo), since there we avoid way more FFI calls.

Also, note that Twitter's DOM is not really huge or deep, which was the worst case for the previous system, and that tests with different count of DOM nodes and different depth, or dirtying DOM elements at the top of the document, could make the improvement way more apparent, while this tests only the hovering over elements almost at the bottom of the DOM.

To summarize, the wins here are two:

 1) Not having to traverse all the node for every tiny incremental restyle, including running flow reconstruction, and
 2) Not having to traverse the whole subtree of an element when marking it as dirty.

I've tried to be fair with the data using a somewhat normal page, but some edge cases could make the servo version without this patch *really* slow.

**Comment by emilio** [comment]
Ok, so I squashed appropriately and I think I got everything right. I need @jdm or @Ms2ger to review e6958d3947adedda7f3f52f26fae321f0b4d0ad4 and 8c48119f7409ef9a122efcc42872cf0577dccab4, since these are load-related changes. I'm leaving some comments on it, but basically is fixing the messages that go to the constellation after triggering DOMContentReady and window load, to do what they are intended to do.

Basically, with e6958d3947adedda7f3f52f26fae321f0b4d0ad4:

 * The `DOMLoad` message to the constellation goes after triggering `DOMContentLoaded`, instead of after triggering the `load` event on the window.
 * The `LoadComplete` message is sent after actually triggering the window's `load` event (and the appropriate reflows), instead of after enqueuing the runnable to do so. This is what was intended according to the documentation in the definition of the message.
 * The `subframe_loaded` is handled after `LoadComplete` instead of `DOMLoad`, which is what was happening before incidentally, since we sent the `DOMLoad` event after `window.onload`.

Also, the other commit, 8c48119f7409ef9a122efcc42872cf0577dccab4, prevents the parent document to get its blocking loads list empty when an iframe navigates to another page. This was the intended thing according to the comment, but in practice I *think* it didn't happen, and now that layout wasn't necessarily busy when the `LoadComplete` message arrived, started to bit in the reftests "is-ready-to-take-an-image detector". I'm trying to verify if that's what happened actually, or if it was one of the above changes the thing that was hitting me. Also, I'm trying to write a test for that, since it *could* have web-visible behavior (I'm thinking of a parent `window.onload` triggered before the child iframe's).

Also, I filled #12591 about the random multiprocess breakage (that for obvious reasons is no longer in this PR).

**Comment by emilio** [comment]
Actually, after looking at it a bit more carefully, I have indeed changed the intended semantics of those messages. I think the actual ones are clearer though. Happy to wait until review.

**Comment by emilio** [comment]
To make a bit clearer what my changes attempted, the constellation used the `LoadComplete` message to identify the document as ready (it was used to take a screenshot later).

The problem is the message is sent before the reflow associated with a document load happens. Before, layout was still busy, so the reftests screenshot wasn't taken until it finished, but now it was being taken.

That makes me wonder:

@jdm: what's the reason for the differences between `DOMLoad` and `LoadComplete` messages to the constellation? 

Sounds to me like before a `LoadComplete` arriving implied a `DOMLoad` coming soon. And what we did in `LoadComplete` was somewhat broken for the reasons stated above.

With this patch `DOMLoad` means `DOMContentReady`, but right now it's only used for webdriver stuff that in fact is intended to run in `LoadComplete`, so after fixing it, I don't think we should need two separate messages.

File: components/style/parallel.rs [hunk 1]
```diff
@@ -63,25 +63,34 @@ fn top_down_dom<N, C>(unsafe_nodes: UnsafeNodeList,
         // Get a real layout node.
         let node = unsafe { N::from_unsafe(&unsafe_node) };
 
+        if !context.should_process(node) {
+            continue;
+        }
```
Comment (bholley) [inline components/style/parallel.rs]: Check the root in traverse_dom and then remove this check?

File: components/style/traversal.rs [hunk 2]
```diff
@@ -147,6 +147,12 @@ pub trait DomTraversalContext<N: TNode>  {
     fn process_preorder(&self, node: N);
     /// Process `node` on the way up, after its children have been processed.
     fn process_postorder(&self, node: N);
+
+    /// Returns if the node should be processed by the preorder traversal.
+    fn should_process(&self, _node: N) -> bool { true }
+
+    /// Do an action over the child before pushing him to the work queue.
```
Comment (bholley) [inline components/style/traversal.rs]: "pushing it"

File: components/style/traversal.rs [hunk 3]
```diff
@@ -245,21 +251,15 @@ pub fn recalc_style_at<'a, N, C>(context: &'a C,
...
-                    }
-                },
-                None => {}
+    if let Some(element) = node.as_element() {
+        if let Some(ref property_declaration_block) = *element.style_attribute() {
+            if property_declaration_block.declarations().any(|d| d.0.has_viewport_percentage()) {
+                unsafe {
+                    node.set_dirty_on_viewport_size_changed();
+                }
+                node.set_descendants_dirty_on_viewport_size_changed();
```
Comment (bholley) [inline components/style/traversal.rs]: This shouldn't be needed?

File: ports/geckolib/traversal.rs [hunk 4]
```diff
@@ -35,5 +36,20 @@ impl<'lc, 'ln> DomTraversalContext<GeckoNode<'ln>> for RecalcStyleOnly<'lc> {
         recalc_style_at(&self.context, self.root, node);
     }
 
+    fn should_process(&self, node: GeckoNode<'ln>) -> bool {
+        node.is_dirty() || node.has_dirty_descendants()
```
Comment (bholley) [inline ports/geckolib/traversal.rs]: File a followup to coalesce this to one FFI call?

**Comment by bholley** [comment]
First three commits lgtm - have someone look at the dom pieces, and have @pcwalton do a high-level signoff on the parallel traversal piece. Nice find!

File: components/style/parallel.rs [hunk 5]
```diff
@@ -63,25 +63,34 @@ fn top_down_dom<N, C>(unsafe_nodes: UnsafeNodeList,
...
+
         // Perform the appropriate traversal.
         context.process_preorder(node);
 
-        let child_count = node.children_count();
+        // Possibly enqueue the children.
+        let mut children_to_process = 0isize;
+        for kid in node.children() {
+            context.pre_process_child_hook(node, kid);
```
Comment (pcwalton) [inline components/style/parallel.rs]: Could you add a comment here explaining what `pre_process_child_hook` typically does? It'll be important for anyone reading this code to understand the flow of dirty bits.

**Comment by pcwalton** [comment]
The approach seems good to me. 👍

**Comment by jdm** [comment]
Everything besides the iframe changes look good to me; those are unnecessary because we check for any outstanding loads before queuing the final document load event.

**Comment by emilio** [comment]
@bors-servo: r=bholley,jdm,pcwalton

Removed the iframe commit, thanks Josh!

**Comment by bholley** [comment]
Tests with unexpected results:
  ▶ Unexpected subtest result in /html/browsers/history/the-history-interface/001.html:
  │ FAIL [expected PASS] pushState should not actually load the new URL
  │   → assert_true: expected true got undefined
  │ 
  │ tests12/<@http://web-platform.test:8000/html/browsers/history/the-history-interface/001.html:283:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1402:20
  │ test@http://web-platform.test:8000/resources/testharness.js:500:9
  └ tests12@http://web-platform.test:8000/html/browsers/history/the-history-interface/001.html:282:9

#12580 #12625

**Comment by emilio** [comment]
@bors-servo: retry

 * #12625 (the other test).
